src/utils.js

```diff
@@ -0,0 +1,94 @@
+import { SHA256 } from './crypto-js/sha256.js';
+import { Base64 } from './crypto-js/enc-base64.js';
+
+function getUrlParms(url){
+    if (!url){
+        url = window.location.hash;
+    }
+
+    if (!url){
+        return {}
+    }
+    const [hash, query] = url.split('#')[1].split('?');
+    const urlParams = Object.fromEntries(new URLSearchParams(hash));
+
+    return urlParams;
+}
+
+function base64URL (string) {
+    return string.toString(Base64)
+    .replace(/=/g, '')
+    .replace(/\+/g, '-')
+    .replace(/\//g, '_');
+}
+
+function generateRandomString(length) {
```

Comment:
Since this string is being used for security purposes, we should find a way to generate a cryptographically secure random string.  See here for an idea

https://security.stackexchange.com/questions/261770/create-a-cryptographically-secure-random-string-for-auth-cookie-in-javascript

Reply:
I researched a little bit into this and found the following.

https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues#usage_notes

So it the official Mozilla documentation, they do not recommend to use `getRandomValues` from `crypto` to generate encryption keys. but they do recommend using instead `generateKey` from `subtleCrypto`.

So I digged more into this and found the following.

https://stackoverflow.com/questions/63309409/creating-a-code-verifier-and-challenge-for-pkce-auth-on-spotify-api-in-reactjs

These are some snippets to generate all the cryptographically secured tokens, and `getRandomValues` is only used for the code verifier token. 

So I will refactor this as well to test it against the identity provider test server.

Reply:
In order to fix the imports and this token generation, I decided we should have our own implementation of the PKCE module/class with its own token generation functions that use the browser built-in crypto module, that way we understand what occurs and at the same time generate cryptographically secured tokens